client/client-multi/client-cpp/impl/utils/ThreadPool.cpp

```diff
@@ -16,128 +16,150 @@
 
 #include "kaa/utils/ThreadPool.hpp"
 
-#include "kaa/logging/Log.hpp"
-#include "kaa/common/exception/KaaException.hpp"
-
+#include <chrono>
+#include <stdexcept>
 
 namespace kaa {
 
-class Worker {
-public:
-    Worker(ThreadPool& threadPool) : threadPool_(threadPool) {}
-    void operator() ();
-
-private:
-    ThreadPool& threadPool_;
-};
-
-void Worker::operator ()()
-{
-    KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPool_.threadPoolGuard_);
-
-    while (threadPool_.isRun_) {
-        while (threadPool_.isRun_ && threadPool_.tasks_.empty() && !threadPool_.isPendingShutdown_) {
-            KAA_CONDITION_WAIT(threadPool_.onNewTask_, tasksLock);
-        }
-
-        if (!threadPool_.isRun_ || (threadPool_.isPendingShutdown_ && threadPool_.tasks_.empty())) {
-            return;
-        }
-
-        auto task = threadPool_.tasks_.front();
-        threadPool_.tasks_.pop_front();
-
-        KAA_UNLOCK(tasksLock);
-
-        try {
-            task();
-        } catch (...) {}
-
-        KAA_LOCK(tasksLock);
-    }
-}
-
-ThreadPool::ThreadPool(std::size_t workerCount): workerCount_(workerCount)
+ThreadPool::ThreadPool(std::size_t workerCount)
+    : workerCount_(workerCount)
 {
     if (!workerCount_) {
-        throw KaaException(boost::format("Failed to create thread pool with %u workers ") % workerCount_);
+        throw std::invalid_argument((boost::format("Wrong thread pool worker count %u") % workerCount_).str());
     }
 }
 
 ThreadPool::~ThreadPool()
 {
-    stop(!shutdownTimer_);
+    shutdownNow();
 }
 
 void ThreadPool::add(const ThreadPoolTask& task)
 {
     if (!task) {
-        throw KaaException("Failed to add task to thread pool: empty callback");
+        throw std::invalid_argument("Null thread pool task");
     }
-    KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPoolGuard_);
 
-    if (isPendingShutdown_) {
-        throw KaaException("Failed to add task to thread pool: pending shutdown");
-    }
+    {
+        KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPoolGuard_);
 
-    if (workers_.empty()) {
-        start();
-    }
+        if (state_ == State::CREATED) {
+            start();
+        }
 
-    tasks_.push_back(task);
+        if (state_ != State::RUNNING) {
+            throw std::runtime_error("Thread pool pending shutdown");
+        }
 
-    KAA_UNLOCK(tasksLock);
+        tasks_.push_back(task);
+    }
 
-    onNewTask_.notify_one();
+    onThreadpoolEvent_.notify_one();
 }
 
 void ThreadPool::awaitTermination(std::size_t seconds)
 {
-    KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPoolGuard_);
+    {
+        KAA_MUTEX_UNIQUE_DECLARE(waitLock, threadPoolGuard_);
 
-    if (isRun_) {
-        shutdownTimer_.reset(new KaaTimer<void()>("Thread pool shutdown timer"));
-        shutdownTimer_->start(seconds, [this] () { stop(true); } );
+        if (state_ != State::PENDING_SHUTDOWN) {
```

Comment:
Maybe it's better to call `shutdown()` here. (Or at least specify `@throw` in the doxygen comment.)